### DISCUSSION TITLE
Help with compiling static libraries for vips-ffm on macOS

### DISCUSSION THREAD

**Discussion by StefanOltmann** [libvips/libvips#4267] [Q&A] [answered]
Hello,

I’m looking to integrate [vips-ffm](https://github.com/lopcode/vips-ffm) into my JVM application.

On Windows, I’ve successfully used the static pre-built DLLs (`libvips-42.dll`, `libglib-2.0-0.dll` and `libgobject-2.0-0.dll`). However, on macOS, I’m struggling to obtain static binaries for both `x86_64` and `aarch64` architectures.

The libraries installed via Homebrew are dynamic, so it seems I’ll need to compile static libraries myself. Unfortunately, I have no prior experience with this process.

I tried following the [libvips cheatsheet](https://github.com/libvips/libvips?tab=readme-ov-file#cheatsheet) and even sought guidance from ChatGPT, but I haven’t been able to generate usable files. The static libraries need to support JPEG, EXIF, JXL, PNG, TIFF and WebP.

Could someone provide step-by-step guidance on how to compile static libraries for macOS?

Kind regards,

Stefan

**Comment by jcupitt** [comment]
Hello @StefanOltmann,

I tried here. First I installed vips from homebrew:

```
brew update && brew upgrade
brew install vips
```

It's an easy way to install all the libvips dependencies. Then I built git master libvips to a non-homebrew prefix. 8.16 will work fine too.

```
cd GIT/libvips
meson setup build --prefix ~/vips --default-library static -Dmodules=disabled
cd build/
ninja install
```

And now I have a static libvips:

```
✔ ~/GIT/libvips [master|✔] 
15:55 $ ls ~/vips/lib
libvips-cpp.a     libvips.a         pkgconfig         
```

It'll still link dynamically against any dependencies which don't have static libraries available, of course, but it might be enough.

Are you certain you need a static library? I would use a dynamic library if possible.

**Reply by StefanOltmann** [reply]
Hi,

thank you you for your quick response.

> Are you certain you need a static library? I would use a dynamic library if possible.

I think so, because I distribute my app to macOS machines that might have non of that pre-installed. So to make sure I can use it there I guess it needs to be static.

Will the libvips.a contain also what it needs from `libglib-2.0-0` &  `libgobject-2.0-0` or do I need to figure out how to build these too?

I assume the `libvips.a` build on a M1 Mac will be aarch64 and for Intel I need to change something, right? I don't have an Intel Mac.

**Reply by jcupitt** [reply]
This this an app as in a `xxxx.app` folder? You don't need static libraries for this -- just make a `lib` subdirectory, put copies of all the libraries in that, and set `DYLD_SEARCH_PATH` in your wrapper script.

I think homebrew makes ARM-only binaries by default now, but they might be mixed. I'd check their docs.

**Reply by jcupitt** [reply]
I made a dynamically linked macOS build of the libvips GUI:

https://github.com/libvips/nip2/releases/download/v8.6.0-beta2/nip2-8.6.0-beta2.app.dmg

It's so old now it probably no longer works, but the wrapper script, `Info.plist`, and directory structure might be useful.

**Reply by StefanOltmann** [reply]
I followed the steps, but trying to load the resulting library in Java using `System.load("libvips.a")` results in `libvips.a (not a mach-o file)`.

I believe that I would need a `libvips.dylib` file.

It looks like that the other dependencies with libraries are also dylib files.
So I guess I'm asking for a single `libvips.dylib` that despite being a "dy(namic)lib" contains all dependencies in one file.

**Reply by jcupitt** [reply, marked as answer]
Exactly, you can't load static libraries at runtime, they are only for compiling.

I would use dynamic libraries if possible.

**Reply by jcupitt** [reply]
... and have a `lib` folder in your app that contains all the dynamic libraries that you need.

**Reply by StefanOltmann** [reply]
Okay, thank you. I'll work on getting everything I need prebuilt from Homebrew and hope it's manageable.

**Comment by StefanOltmann** [comment]
Unfortunately, I need more assistance here. :/

I managed to build `libvips.42.dylib` using the command `meson setup build --prefix ~/vips --default-library static -Dmodules=disabled`, and it worked fine.

However, the `libglib-2.0.0.dylib` and `libgobject-2.0.0.dylib` libraries I copied from `/opt/homebrew/lib` depend on other libraries in that folder, and the paths seem to be hardcoded.

What I need is to compile `libvips.42.dylib`, `libglib-2.0.0.dylib`, and `libgobject-2.0.0.dylib` in a way that they look for each other in the same directory. Additionally, I may need to include other dependent libraries.

Is it possible to build `libvips.42.dylib` in a way that it contains all of its necessary dependencies?

**Reply by jcupitt** [reply]
Did you try setting `DYLD_LIBRARY_PATH`? I think that should fix it, though it's been a while since I tried it myself, perhaps things have changed.

**Reply by StefanOltmann** [reply]
The problem is that I want to distribute libvips bundles with my app to systems that don’t have the dependencies installed. I need to deliver everything and if libraries look for absolute paths on the system they were build on this won’t work.

So I’m looking for a way to have only three dylibs that contain everything.

On Windows I use the prebuilt DLLs which don’t have further dependencies.

There is nothing like that for macOS.

**Reply by jcupitt** [reply]
I understand, but `DYLD_LIBRARY_PATH` will override the path built into the library. That's how the nip2.app I linked above works.

(or that's how it works on linux and that how it used to work on macOS, I've not tried recently, I should check)

**Reply by StefanOltmann** [reply]
Thank you for your help! 😊

VIPS is 4x faster at generating thumbnails compared to Skiko (the SKIA wrapper for Kotlin)! 🚀 Truly awesome work you've done here.

I’m surprised I didn’t discover this library sooner, considering how long it’s been available.

VIPS will be used in the next version of [Ashampoo Photo Organizer](https://www.ashampoo.com/photo-organizer).

If you’re interested, I’d be happy to gift you a PRO license. :)

**Reply by jcupitt** [reply]
Great! I'm glad it's working.

You should find memory use is quite a bit lower too -- you ought to be able to run many thumbnail tasks in parallel without filling up RAM. This can give a very useful speedup on most modern CPUs.

For example, I made a directory of 1000 6k x 4k JPGs:

```
john@banana ~/pics/sample $ ls
john@banana ~/pics/sample $ for i in {1..1000}; do cp ../nina.jpg $i.jpg; done
john@banana ~/pics/sample $ 
```

Then thumbnail them all to fit 128 x 128:

```
john@banana ~/pics/sample $ /usr/bin/time -f %M:%e vipsthumbnail *.jpg 
60780:26.84
john@banana ~/pics/sample $ 
```

That's 60mb of memory at peak, 26s of runtime, so 26ms per image.

With GNU parallel I see:

```
john@banana ~/pics/sample $ /usr/bin/time -f %M:%e parallel vipsthumbnail {} ::: *.jpg 
40932:3.82
john@banana ~/pics/sample $ 
```

The 40mb is for the controlling process, but it'll be about 600mb in total. The runtime is 3.8s, or 4ms per image.

Another option of course is not to generate thumbnails in advance. libvips is fast enough (usually) that you can make thumbnails on demand and the user won't notice. Most of the web's image resize proxies use libvips like this, and the cost, storage and quality (since images are always sized correctly for the user's device) savings are considerable.

**Reply by jcupitt** [reply]
Your photo organiser looks nice! Though I don't need one myself.

I would look at building libvips from source and not including formats you don't support. By default it includes strange things (eg. Matlab) which will probably not be useful to your customers and which we cannot test for security. 

Failing that, you can set the env var `VIPS_BLOCK_UNTRUSTED` which will disable loaders we don't fuzz at runtime.

libvips won't be able to generate thumbnails for PNG on demand (the PNG format is not designed for speed), so you'd need to at least pre-generate derivatives for that.

**Reply by StefanOltmann** [reply]
Thank you for all your hints! :)
